src/parser/jobs/ast/BuffExtensions.js

```diff
@@ -0,0 +1,270 @@
+
+
+import BuffList from './BuffList'
+import styles from './BuffExtensions.module.css'
+
+// TODO: Make some inference on their CO and TD usage for the suggestions panel - Sushi
+export default class BuffExtensions extends Module {
+	static handle = 'buffextensions'
+	static title = 'Buff Extensions'
+	static dependencies = [
+		'suggestions',
+	]
+
+	// Latency allowance for applying aoe effects
+	_envEffectGracePeriod = 4500
+	// Array of objects detailing each use of either Time Dilation or Celestial Opposition
+	_dilationUses = []
+	_oppositionEvent = null
+
+
+	constructor(...args) {
+		super(...args)
+
+		const filter = {
+			by: 'player',
+			abilityId: [ACTIONS.TIME_DILATION.id, ACTIONS.CELESTIAL_OPPOSITION.id],
+		}
+
+		this.addHook('cast', filter, this._onCast)
+		this.addHook('refreshbuff', {by: 'player'}, this._onBuffRefresh)
+
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCast(event) {
+		const actionID = event.ability.guid
+
+		if (actionID === ACTIONS.TIME_DILATION.id) {
+			this._onDilation(event)
+		}
+
+		if (actionID === ACTIONS.CELESTIAL_OPPOSITION.id) {
+			this._onOpposition(event)
+		}
+
+
+	}
+
+	/**
+	 * Grabs all the buffs the target has, packs it up with some info and add to _dilationUses
+	 *
+	 * @param {obj} log event of a Time Dilation cast
+	 * @return {void} null
+	 */
+	_onDilation(event) {
+
+		// TODO: prevent getEntity from returning null if target was a pet - Sushi
+		// (Pets are stored under the player entities)
+		const refreshedTarget = this.parser.modules.combatants.getEntity(event.targetID)
+
+		this._dilationUses.push({
+			event: event,
+			targets: [{
+				id: event.targetID,
+				name: refreshedTarget.info.name,
+				job: refreshedTarget.info.type,
+				buffs: refreshedTarget.getStatuses(event.timestamp, 0, 1000, event.sourceID),
+			}],
+
+		})
+	}
+
+	/**
+	 * Prepares an object to be populated by refresh events
+	 *
+	 * @param {obj} log event of a Celestial Opposition cast
+	 * @return {void} null
+	 */
+	_onOpposition(event) {
+
+		// Clear previous Opposition history
+		if (this._oppositionEvent) {
+			this._dilationUses.push({...this._oppositionEvent})
+			this._oppositionEvent = null
+		}
+
+		// Structure a new collection of refreshed buffs by this Opposition cast
+		this._oppositionEvent = {
+			event: event,
+			targets: [],
+		}
+
+	}
+
+
+	/**
+	 * Checks if the timestamp is within allowable range from the Opposition Event
+	 * Populates the _oppositionEvent created in _onOpposition with refresh events on friendlies
+	 *
+	 * @param {obj} log event of a Celestial Opposition cast
+	 * @return {void} null
+	 */
+	_onBuffRefresh(event) {
+		const statusID = event.ability.guid
+
+		// Ignore if timestamp is after aoe effect grace period
+		if (this._oppositionEvent && event.timestamp < (this._oppositionEvent.event.timestamp + this._envEffectGracePeriod)) {
+
+			// Ignore refreshes on protects, royal road statuses,
+			if (statusID !== STATUSES.PROTECT.id
```

Comment:
You can consolidate this up top of the file with something like

```js
const IGNORE_STATUSES = [
	STATUSES.PROTECT.id,
	STATUSES.COLLECTIVE_UNCONSCIOUS.id,
	...ROYAL_ROAD_STATES,
	...HELD_ARCANA,
	...DRAWN_ARCANA,
]
```

That way you only need to do the one `IGNORE_STATUSES.includes(statusId)` down here.